linfa-clustering/src/gaussian_mixture/algorithm.rs

```diff
@@ -0,0 +1,494 @@
+/// let dataset = Dataset::from(generate_blobs(n, &expected_centroids, &mut rng));
+///
+/// // Our GMM is expected to have a number of clusters equals the number of centroids
+/// // used to generate the dataset
+/// let n_clusters = expected_centroids.len_of(Axis(0));
+///
+/// // We fit the model from the dataset setting some options
+/// let gmm = GaussianMixtureModel::params_with_rng(n_clusters, rng)
+///             .n_init(10)
+///             .tolerance(1e-4)
+///             .build()
+///             .fit(&dataset);
+///
+/// // We can get predicted centroids (ie means of learnt gaussian distributions) from the model
+/// let gmm_centroids = gmm.centroids();
+///
+/// // We can check that centroids used to generate test dataset are close to GMM centroids
+/// let memberships = gmm.predict(&expected_centroids);
+/// for (i, expected_c) in expected_centroids.outer_iter().enumerate() {
+/// let closest_c = gmm_centroids.index_axis(Axis(0), memberships[i]);
+/// Zip::from(&closest_c)
+///     .and(&expected_c)
+///     .apply(|a, b| assert_abs_diff_eq!(a, b, epsilon = 1.))
+/// }
+/// ```
+pub struct GaussianMixtureModel<F: Float> {
+    covar_type: GmmCovarType,
+    weights: Array1<F>,
+    means: Array2<F>,
+    precisions: Array3<F>,
+}
+
+impl<F: Float> Clone for GaussianMixtureModel<F> {
+    fn clone(&self) -> Self {
+        Self {
+            covar_type: self.covar_type,
+            weights: self.weights.to_owned(),
+            means: self.means.to_owned(),
+            precisions: self.precisions.to_owned(),
+        }
+    }
+}
+
+impl<F: Float + Into<f64>> GaussianMixtureModel<F> {
+    pub fn params(n_clusters: usize) -> GmmHyperParamsBuilder<F, Isaac64Rng> {
+        GmmHyperParams::new(n_clusters)
+    }
+
+    pub fn params_with_rng<R: Rng + Clone>(
+        n_clusters: usize,
+        rng: R,
+    ) -> GmmHyperParamsBuilder<F, R> {
+        GmmHyperParams::new_with_rng(n_clusters, rng)
+    }
+
+    pub fn weights(&self) -> &Array1<F> {
+        &self.weights
+    }
+
+    pub fn means(&self) -> &Array2<F> {
+        &self.means
+    }
+
+    pub fn precisions(&self) -> &Array3<F> {
+        &self.precisions
+    }
+
+    pub fn centroids(&self) -> &Array2<F> {
+        self.means()
+    }
+
+    fn new<D: Data<Elem = F>, R: Rng + Clone, T: Targets>(
+        hyperparameters: &GmmHyperParams<F, R>,
+        dataset: &Dataset<ArrayBase<D, Ix2>, T>,
+        mut rng: R,
+    ) -> GaussianMixtureModel<F> {
+        let observations = dataset.records().view();
+        let n_samples = observations.nrows();
+
+        let resp = match hyperparameters.init_method() {
+            GmmInitMethod::KMeans => {
+                let model = KMeans::params_with_rng(hyperparameters.n_clusters(), rng)
+                    .build()
+                    .fit(&dataset);
+                let mut resp = Array::<F, Ix2>::zeros((n_samples, hyperparameters.n_clusters()));
```

Comment:
what does `resp` mean here, responsibility?

Reply:
yes, resp is for responsibility ie the probability a sample belongs to a given cluster. I will add a comment.